pkg/virt-controller/watch/replicaset.go

```diff
@@ -0,0 +1,357 @@
+		return false
+	}
+	defer c.queue.Done(key)
+	if err := c.execute(key.(string)); err != nil {
+		logging.DefaultLogger().Info().Reason(err).Msgf("reenqueuing VirtualMachineReplicaSet %v", key)
+		c.queue.AddRateLimited(key)
+	} else {
+		logging.DefaultLogger().Info().V(4).Msgf("processed VirtualMachineReplicaSet %v", key)
+		c.queue.Forget(key)
+	}
+	return true
+}
+
+func (c *VMReplicaSet) execute(key string) error {
+
+	obj, exists, err := c.vmRSInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return nil
+	}
+	if !exists {
+		// nothing we need to do. It should always be possible to re-create this type of controller
+		return nil
+	}
+	rs := obj.(*virtv1.VirtualMachineReplicaSet)
+
+	//TODO default rs if necessary, the aggregated apiserver will do that in the future
+	if rs.Spec.Template == nil || rs.Spec.Selector == nil || rs.Spec.Selector.Size() == 0 {
+		logging.DefaultLogger().Object(rs).Error().Msg("Invalid controller spec, will not retry processing it.")
+		return nil
+	}
+
+	selector, err := v1.LabelSelectorAsSelector(rs.Spec.Selector)
+	if err != nil {
+		return nil
+	}
+
+	// get all potentially interesting VMs from the cache
+	vms, err := c.listVMsFromNamespace(rs.ObjectMeta.Namespace)
+
+	if err != nil {
+		return err
+	}
+
+	// make sure we only consider active VMs
+	vms = c.filterActiveVMs(vms)
+
+	// make sure the selector of the controller matches and the VMs match
+	vms = c.filterMatchingVMs(selector, vms)
+
+	// Scale up or down
+	scaleErr := c.scale(rs, vms)
+
+	obj, err = model.Clone(rs)
+	if err != nil {
+		return err
+	}
+	rsCopy := obj.(*virtv1.VirtualMachineReplicaSet)
+
+	return c.updateStatus(rsCopy, vms, scaleErr)
+}
+
+func (c *VMReplicaSet) scale(rs *virtv1.VirtualMachineReplicaSet, vms []virtv1.VM) error {
+
+	diff := c.calcDiff(rs, vms)
+
+	if diff == 0 {
+		return nil
+	}
+
+	// Every delete request can fail, give the channel enough room, to not block the go routines
+	errChan := make(chan error, abs(diff))
+
+	var wg sync.WaitGroup
+	wg.Add(abs(diff))
+
+	if diff > 0 {
+		// We have to delete VMs
+		for i := 0; i < diff; i++ {
+			go func(idx int) {
+				defer wg.Done()
+				deleteCandidate := vms[idx]
+				// TODO graceful delete
+				err := c.clientset.VM(rs.ObjectMeta.Namespace).Delete(deleteCandidate.ObjectMeta.Name, &v1.DeleteOptions{})
+				if err != nil {
+					errChan <- err
+				}
+			}(i)
+		}
+
+	} else if diff < 0 {
+		// We have to create VMs
+		for i := diff; i < 0; i++ {
+			go func() {
+				defer wg.Done()
+				vm := virtv1.NewVMReferenceFromNameWithNS(rs.ObjectMeta.Namespace, "")
+				vm.ObjectMeta.GenerateName = rs.ObjectMeta.Name + "-"
+				vm.Spec = rs.Spec.Template.Spec
+				// TODO check if vm labels exist, and when make sure that they match. For now just override them
+				vm.ObjectMeta.Labels = rs.Spec.Template.ObjectMeta.Labels
+				_, err := c.clientset.VM(rs.ObjectMeta.Namespace).Create(vm)
+				if err != nil {
+					errChan <- err
+				}
+			}()
+		}
+	}
+	wg.Wait()
```

Comment:
So this blocks until the changes to VM count have been scheduled?

Reply:
It could (I think the 8s ReplicaSet does), here I just do concurrent request on the apiserver, to not add up request delays, so that we can release the controller worker faster. I kind of borrowed that from the ReplicaSet controller from k8s.